server/lib/whispers/whisper-api.ts

```diff
@@ -0,0 +1,171 @@
+import Router, { RouterContext } from '@koa/router'
+import httpErrors from 'http-errors'
+import Joi from 'joi'
+import Koa from 'koa'
+import { container } from 'tsyringe'
+import { assertUnreachable } from '../../../common/assert-unreachable'
+import { USERNAME_MAXLENGTH, USERNAME_MINLENGTH, USERNAME_PATTERN } from '../../../common/constants'
+import { httpApi, HttpApi } from '../http/http-api'
+import ensureLoggedIn from '../session/ensure-logged-in'
+import createThrottle from '../throttle/create-throttle'
+import throttleMiddleware from '../throttle/middleware'
+import { validateRequest } from '../validation/joi-validator'
+import WhisperService, { WhisperServiceError, WhisperServiceErrorCode } from './whisper-service'
+
+const startThrottle = createThrottle('whisperstart', {
+  rate: 3,
+  burst: 10,
+  window: 60000,
+})
+
+const closeThrottle = createThrottle('whisperclose', {
+  rate: 10,
+  burst: 20,
+  window: 60000,
+})
+
+const sendThrottle = createThrottle('whispersend', {
+  rate: 30,
+  burst: 90,
+  window: 60000,
+})
+
+const retrievalThrottle = createThrottle('whisperretrieval', {
+  rate: 30,
+  burst: 120,
+  window: 60000,
+})
+
+function isWhisperServiceError(error: Error): error is WhisperServiceError {
+  return error.hasOwnProperty('code')
+}
+
+function convertWhisperServiceError(err: Error) {
+  if (!isWhisperServiceError(err)) {
+    throw err
+  }
+
+  switch (err.code) {
+    case WhisperServiceErrorCode.UserOffline:
+    case WhisperServiceErrorCode.UserNotFound:
+      throw new httpErrors.NotFound(err.message)
+    case WhisperServiceErrorCode.InvalidCloseAction:
+    case WhisperServiceErrorCode.InvalidGetSessionHistoryAction:
+      throw new httpErrors.BadRequest(err.message)
+    case WhisperServiceErrorCode.NoSelfMessaging:
+      throw new httpErrors.Forbidden(err.message)
+    default:
+      assertUnreachable(err.code)
```

Comment:
Didn't really think about this in the other places you've done this, but this isn't *really* safe because `code` doesn't necessarily guarantee that the thing we have here is actually a WhisperServiceError. Might be good to add a Symbol or something on the prototype that you can check in `isWhisperServiceError` instead, and then this assertion wouldn't be replacing e.g. errors from `fs`.

(Alternative to the Symbol approach if you know the classes will always be constructed WhisperServiceErrors, which seems reasonable, is simply to do an `instanceof WhisperServiceError` check)